File: 17.js
/**
 * @param {string} digits
 * @return {string[]}
 */
// 2=a,b,c
// 3=d,e,f
// r=""
// 1="a" "b"
// 2="ad" "ae" "af" "b"

/*
a
ad, 
*/
var letterCombinations = function(digits) {
    let res = [];
    let dictionary = {
        2: ["a", "b", "c"],
        3: ["d", "e", "f"],
        4: ["g", "h", "i"],
        5: ["j", "k", "l"],
        6: ["m", "n", "o"],
        7: ["p", "q", "r", "s"],
        8: ["t", "u", "v"],
        9: ["w", "x", "y", "z"]
    }
    
    if(digits == ""){
        return [];
    }
    function recursive(string, digits, depth) {
        if(depth === digits.length) {
            res.push(string);
        } else {
            for(let i=0;i<dictionary[digits[depth]].length;i++) {
                string = string.concat(dictionary[digits[depth]][i]);
                recursive(string, digits, depth+1);
                string = string.substring(0, string.length - 1);
            }
        }   
    }
    
    recursive("", digits, 0);
    return res;
};

File: 17.java
import java.util.*;

public class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        
        if (digits == null || digits.length() == 0) {
            return res;
        }

        Map<Character, String> dictionary = new HashMap<>();
        dictionary.put('2', "abc");
        dictionary.put('3', "def");
        dictionary.put('4', "ghi");
        dictionary.put('5', "jkl");
        dictionary.put('6', "mno");
        dictionary.put('7', "pqrs");
        dictionary.put('8', "tuv");
        dictionary.put('9', "wxyz");

        recursive(digits, 0, new StringBuilder(), dictionary, res);
        return res;
    }

    private void recursive(String digits, int depth, StringBuilder current, Map<Character, String> dictionary, List<String> res) {
        if (depth == digits.length()) {
            res.add(current.toString());
            return;
        }

        String letters = dictionary.get(digits.charAt(depth));
        for (int i = 0; i < letters.length(); i++) {
            current.append(letters.charAt(i));
            recursive(digits, depth + 1, current, dictionary, res);
            current.deleteCharAt(current.length() - 1);  // Backtrack
        }
    }
}
